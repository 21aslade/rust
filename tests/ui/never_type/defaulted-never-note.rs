//@ revisions: nofallback fallback

// We need to opt into the `never_type_fallback` feature
// to trigger the requirement that this is testing.
#![cfg_attr(fallback, feature(never_type, never_type_fallback))]

#![allow(unused)]

trait Deserialize: Sized {
    fn deserialize() -> Result<Self, String>;
}

impl Deserialize for () {
    fn deserialize() -> Result<(), String> {
        Ok(())
    }
}

trait ImplementedForUnitButNotNever {}

impl ImplementedForUnitButNotNever for () {}

fn foo<T: ImplementedForUnitButNotNever>(_t: T) {}
//[fallback]~^ note: required by this bound in `foo`
//[fallback]~| note: required by a bound in `foo`
fn smeg() {
    //[nofallback]~^ error: this function depends on never type fallback being `()`
    //[nofallback]~| warn: this was previously accepted by the compiler but is being phased out; it will become a hard error in Rust 2024 and in a future release in all editions!
    let _x = return;
    foo(_x);
    //[fallback]~^ error: the trait bound
    //[fallback]~| note: the trait `ImplementedForUnitButNotNever` is not implemented
    //[fallback]~| help: trait `ImplementedForUnitButNotNever` is implemented for `()`
    //[fallback]~| note: this error might have been caused
    //[fallback]~| note: required by a bound introduced by this call
    //[fallback]~| help: you might have intended to use the type `()`
}

fn main() {
    smeg();
}
