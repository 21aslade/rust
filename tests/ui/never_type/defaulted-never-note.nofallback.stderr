error: this function depends on never type fallback being `()`
  --> $DIR/defaulted-never-note.rs:26:1
   |
LL | fn smeg() {
   | ^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in Rust 2024 and in a future release in all editions!
   = note: for more information, see <https://doc.rust-lang.org/edition-guide/rust-2024/never-type-fallback.html>
   = help: specify the types explicitly
note: in edition 2024, the requirement `!: ImplementedForUnitButNotNever` will fail
  --> $DIR/defaulted-never-note.rs:30:9
   |
LL |     foo(_x);
   |         ^^
   = note: `#[deny(dependency_on_unit_never_type_fallback)]` (part of `#[deny(rust_2024_compatibility)]`) on by default
help: use `()` annotations to avoid fallback changes
   |
LL |     let _x: () = return;
   |           ++++

error: aborting due to 1 previous error

Future incompatibility report: Future breakage diagnostic:
error: this function depends on never type fallback being `()`
  --> $DIR/defaulted-never-note.rs:26:1
   |
LL | fn smeg() {
   | ^^^^^^^^^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in Rust 2024 and in a future release in all editions!
   = note: for more information, see <https://doc.rust-lang.org/edition-guide/rust-2024/never-type-fallback.html>
   = help: specify the types explicitly
note: in edition 2024, the requirement `!: ImplementedForUnitButNotNever` will fail
  --> $DIR/defaulted-never-note.rs:30:9
   |
LL |     foo(_x);
   |         ^^
   = note: `#[deny(dependency_on_unit_never_type_fallback)]` (part of `#[deny(rust_2024_compatibility)]`) on by default
help: use `()` annotations to avoid fallback changes
   |
LL |     let _x: () = return;
   |           ++++

